revscoring/features/tests/test_revision.py

```diff
@@ -0,0 +1,292 @@
+    }
+    eq_(solve(words, cache=cache), 4)
+
+def test_badwords():
+    def is_badword(w): return w == "badword"
+    cache = {
+        languages.is_badword: is_badword,
+        revision.words: ["I", "am", "badword", "badword"]
+    }
+    eq_(solve(badwords, cache=cache), 2)
+
+def test_misspellings():
+    def is_misspelled(w): return w == "misspelled"
+    cache = {
+        languages.is_misspelled: is_misspelled,
+        revision.words: ["I", "am", "misspelled", "badword"]
+    }
+    eq_(solve(misspellings, cache=cache), 1)
+
+############################# Parsed text ######################################
+
+def test_level_1_headings():
+    FakeHeading = namedtuple("FakeHeading", ['level'])
+    cache = {
+        revision.headings: [FakeHeading(2), FakeHeading(3), FakeHeading(4),
+                             FakeHeading(2)]
+    }
+    eq_(solve(level_1_headings, cache=cache), 0)
+
+def test_level_2_headings():
+    FakeHeading = namedtuple("FakeHeading", ['level'])
+    cache = {
+        revision.headings: [FakeHeading(2), FakeHeading(3), FakeHeading(4),
+                             FakeHeading(2)]
+    }
+    eq_(solve(level_2_headings, cache=cache), 2)
+
+def test_level_3_headings():
+    FakeHeading = namedtuple("FakeHeading", ['level'])
+    cache = {
+        revision.headings: [FakeHeading(2), FakeHeading(3), FakeHeading(4),
+                             FakeHeading(2)]
+    }
+    eq_(solve(level_3_headings, cache=cache), 1)
+
+def test_level_4_headings():
+    FakeHeading = namedtuple("FakeHeading", ['level'])
+    cache = {
+        revision.headings: [FakeHeading(2), FakeHeading(3), FakeHeading(4),
+                             FakeHeading(2)]
+    }
+    eq_(solve(level_4_headings, cache=cache), 1)
+
+def test_level_5_headings():
+    FakeHeading = namedtuple("FakeHeading", ['level'])
+    cache = {
+        revision.headings: [FakeHeading(2), FakeHeading(3), FakeHeading(4),
+                             FakeHeading(2)]
+    }
+    eq_(solve(level_5_headings, cache=cache), 0)
+
+def test_level_6_headings():
+    FakeHeading = namedtuple("FakeHeading", ['level'])
+    cache = {
+        revision.headings: [FakeHeading(2), FakeHeading(3), FakeHeading(4),
+                             FakeHeading(2)]
+    }
+    eq_(solve(level_6_headings, cache=cache), 0)
+
+def test_infonoise():
+    cache = {
+        languages.is_stopword: lambda w: w == "is",
+        languages.stem_word: lambda w: w[0], # Just the first character
+        revision.content_words: ["This", "is", "twelve"] # 12 characters of content words
+    }
+    eq_(solve(infonoise, cache=cache), 2/12)
+
+def test_category_links():
+    FakeLink = namedtuple("FakeLink", "title")
+    cache = {
+        revision.internal_links: [FakeLink("Foobar"),
+                                  FakeLink("File:Hat.jpg"),
+                                  FakeLink("file:Bum.gif"),
+                                  FakeLink("Category:Hats")]
```

Comment:
Once we have regexes per language, we would need a test for e.g. "Categoria:Matemática" in Portuguese...